docs/adr/adr-11-blocksync-overhaul.md

```diff
@@ -0,0 +1,329 @@
+
+The introduced `EDSStore` also maintains (via DAGStore) a top-level index enabling granular and efficient random access
+to every share and/or Merkle proof over every registered CARv1 file. The `EDSStore` provides a custom `Blockstore` interface
+implementation to achieve the access. The main use-case is randomized sampling over the whole chain of EDS block data
+and getting data by namespace.
+
+EDSStore constructor should instantiate
+
+```go
+type EDSStore struct {
+ basepath string 
+ dgstr dagstore.DAGStore
+ topIdx index.Inverted 
+ carIdx index.FullIndexRepo
+ mounts  *mount.Registry
+ ...
+}
+
+// NewEDSStore constructs EDStore over OS directory to store CARv1 files of EDSes and indices for them.
+// Datastore is used to keep inverted/top-level index.
+func NewEDSStore(basepath string, ds datastore.Batching) *EDSStore {
+ topIdx := index.NewInverted(datastore)
+ carIdx := index.NewFSRepo(basepath + "/index")
+ mounts := mount.NewRegistry()
+ return &EDSStore{
+  basepath: basepath,
+  dgst: dagstore.New(dagstore.Config{...}),
+  topIdx: index.NewInverted(datastore),
+  carIdx: index.NewFSRepo(basepath + "/index")
+  mounts: mounts,
+    }   
+}
+
+```
+
+##### `share.EDSStore.Put`
+
+To write an entire EDS `Put` method is introduced. Internally it
+
+- Opens a file under `storepath/DataRoot` path
+- Serializes the EDS into the file via `share.WriteEDS`
+- Wraps it with `DAGStore`'s [FileMount](https://github.com/filecoin-project/dagstore/blob/master/mount/file.go#L10)
+- Converts `DataRoot` into the [`shard.Key`](https://github.com/filecoin-project/dagstore/blob/master/shard/key.go#L12)
+- Registers the `Mount` as a `Shard` on the `DAGStore`
+  - This register the `Mount` in [`mount.Registry`](https://github.com/filecoin-project/dagstore/blob/master/mount/registry.go#L22)
+passed to `DAGStore`.
+
+NOTE: Registering on the DAGStore automatically populates top-level index with shares/proofs accessible from stored EDS
+so this is out of scope of the document.
+
+```go
+// Put stores the given data square with DataRoot as key.
+//
+// The square is verified on the Exchange level and Put only stores the square trusting it.
+// The resulting file stores all the shares and NMT Merkle Proofs of the EDS.
+// Additionally, the file gets indexed s.t. Store.Blockstore can access them. 
+func (s *Store) Put(context.Context, DataRoot, *rsmt2d.ExtendedDataSquare) error
+```
+
+##### `share.EDSStore.GetCAR`
+
+To read an EDS as a byte stream `GetCAR` method is introduced. Internally it
+
+- Converts `DataRoot` into the [`shard.Key`](<https://github.com/filecoin-project/dagstore/blob/master/shard/key.go#L12>
+- Gets Mount by `shard.Key` from [`mount.Registry`](https://github.com/filecoin-project/dagstore/blob/master/mount/registry.go#L22)
+- Returns `io.ReadeCloser` from [`Mount.Fetch`](https://github.com/filecoin-project/dagstore/blob/master/mount/mount.go#L71)
+
+NOTE: It might be necessary to acquire EDS mount via `DAGStore` keeping `ShardAccessor`
```

Comment:
It's starting to feel uncertain about this when you realize that Acquire does not return you a Reader

Reply:
my POV here is that caching shardaccessors is really nice because it allows you to access the underlying blockstore, which is at least good for bitswap

Reply:
> which is at least good for bitswap 

Yes, and irrelevant for the case above(Get). Get just needs to get a Reader and serve it above. There might be multiple readers accessing the same data. One of those readers might be a DAGStore shardaccessor.

Reply:
Ok, so it's not locking it to one only accessor(it has a refcount), which is good, and we should acquire it then.
(I assumed that acquire locks to one accessor because of it's the word "acquire", which assumes one owner)

Reply:
True, that naming is not that nice